lib/tasks/school_urn_generator.rb

```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+ALL_URNS = (1..999_999).freeze unless defined?(ALL_URNS)
+
+class SchoolURNGenerator
+  class << self
+    def next
+      sprintf("%06d", taken.push(available.pop || (raise "URN available list exhausted"))[-1])
+    end
+
+  private
+
+    def available
+      @available.presence || reseed
+    end
+
+    def reseed
+      @available = (1_000.times.map { rand(ALL_URNS) }.uniq - taken)
+    end
```

Comment:
What is this doing? Why do we need 1000 copies of `rand(ALL_URNS)`?

Reply:
Is this saying that available urns are 1000 random urns, but we remove duplicates and taken? Why do we need this? So the memory doesn't go haywire?

If so, why would we not just use `rand(ALL_URNS)` and re-roll when we run into a taken object? What's the improvement here, other than we do it 1000 times?

Reply:
Ah, I guess if we fail to find a new element in 1000 attempts the code above will fail raise the exception.

Reply:
Yes. That's correct. It's a much lower memory version that create 1000 random URNs, removes duplicate and taken and that becomes the available set for the next set of runs. If it's empty, it raises when next is called.

Reply:
It's a memory optimisation and doing like this means that we don't get any collisions and don't need to cater for them in this.

Reply:
Alright, fair enough